README.txt
Connexions Packaging of CNXML

   CNXML is a lightweight XML markup language designed for marking up
   educational content
   http://cnx.rice.edu/technology/cnxml/

  - This package includes:

    * RelaxNG definition of CNXML 0.7

    * XSLT stylesheets to convert CNXML 0.5 to 0.6, and CNXML 0.6 to 0.7

    * XSL stylesheets to convert to HTML including CALS table support
      from the DocBook XSL Stylesheets

    * XML and SGML Catalog files

For external developers:

    * Developers on systems other than Debian derivates should use 
      the cnxml-jing.rng schema files, which point to component 
      schema files with relative paths on the local filesystem
    * Debian developers can build .deb packages of CNXML with 
      dpkg-buildpackage – installing the .debs will update the XML
      catalog with schema file locations.
    
To run the tests:

    * ./test/vtest.sh [schema-file]
      Runs on every file whose name ends with the '.xml' extension in sub- and sub-subdirectories of the working directory, and prints all error messages to stdout
      test/test.out contains the output of the tests using the rng file in SVN
      Example: ./test/vtest.sh ./schema/cnxml-jing.rng > ./test/test.out && svn st ./test/test.out


CHANGES.txt
cnxml-0.7
  - collection-structure-redesign-1 (Oct. 22, 2009)
    - permit more than two media object children of 'media', with the 'for' 
      attribute to indicate the output medium for which the media object is 
      intended ('online' and 'pdf' supported at present)
    - add MDML version number to 'metadata' element
    - use new MDML version to match ColXML
    - begin integrating language documentation in the schema
    - add test routine and update test cases for new language version 
    - improve RNG simplificator.xsl so that it produces a monolithic schema that validates correctly
    - create separate schema for validating fragments for EIP and abstracts
    - add stylesheet for converting CNXML 0.6 to 0.7
    - LICENSE.txt added
    

cnxml-0.6.3
  - Module Edit Workflow (DATE)
    - #4266: Output empty links to bibtex entries as bracketed numbers.

cnxml-0.6.2
  - Modify localization system to improve performance ("gentext-be-gone", 4/16/09):
    - create a version of the Docbook gentext elements containing
      only the keys that we need;
    - implement a reduced version of the Docbook 'gentext' template;
    - add scripts to compute all gentext keys needed for a given set 
      of module_export_template files, and to create a local l10n.xml
      file from the Docbook files with only the needed keys.

cnxml-0.6.1
  - Put the OER icon, message, and links in the footer (ticket:8592)

cnxml-0.6.0.3
  - Mathml warning message for unsupported browsers

cnxml-0.6.0.2
  - cnxml-0.6-hot-fix, third set (March 6, 2009)
    - make <audio> implement title <param>s in the same way as 0.5 audio media
    - properly implemenet <java-applet> (@code instead of @src) (ticket:8452)
    - change output of <section> <label>s to match other <label> outputs
    - add curly quotes around inline <quote>, unless it has a "no-marks" class
      (which upconverted content will) (ticket:8256)
    - don't add IDs to inline <quote> in upconversion (ticket:8058)
    - fix problem with <preformat> rendering messing up EIP (ticket:8262)
    - preserve <cnxn> ID in upconversion (ticket:8214) and drop "cnxn" class

cnxml-0.6.0.1
  - cnxml-0.6-hot-fix, first set: Upconversion tweaks and bugfixes
    - fix rule and note @types (#8213)
    - make text/@mime-type optional (#8176)
    - eliminate class="cnxn" from upgrade output

cnxml-0.6 (p1)
  - Major release: new version of CNXML. Changes include:
    - implement canonical schema language in Relax NG instead of in DTD;
    - provide a single schema for all CNXML 0.6, including support for MathML and QML;
    - made content models of most elements more generic;
    - add optional 'class' attribute to all elements;
    - add elements 'div', 'span', 'preformat', 'label', 'sub', 'sup', 'newline', 'space';
    - restructure 'media' element and add child elements for types of media: 'image', 'audio', 'video', 'labview', 'flash', 'java-applet', 'download', 'object';
    - enhance author control over appearance of lists;
    - give authors control over placement of exercise solutions;
    - require 'id' attribute on more elements;
    - accessibility enhancements:
      - add 'alt' attribute on 'media' elements;
      - add 'summary' attribute on 'table' elements;
      - support 'rowheader' class on table cells;
    - provide XSLT transformation to convert CNXML 0.5 modules to CNXML 0.6;
    - generalize linking attributes on elements that entail hyperlinks;
    - permit any namespaced attribute (except those in the CNXML namespace) on any element;
    - give authors more control over the numbering of elements;
    - let authors specify typeface attributes (e.g. italics, bold) in 'emphasis' and 'span';

cnxml-0.5 (p24)
  - quickperf (30 Sept 2008)
    - two entries for "Recently" and "Viewed" due to wrapping issues in portlet title

cnxml-0.5 (p23)
  - MyCNX phase 2
    - #6578: duplication problem with translation of "cite"
    - #6275: Label and number exercises (and the empty cnxns to them)

CNXML-0.5 (p22.1)
  - #6531: remove some accidental debugging text

CNXML-0.5 (p22)
  - MyCNX lands
    - unibrowser solution/exercise improvements
    - l10n changes for content_render related to new content actions portlet, favorites, recently viewed portlet

CNXML-0.5 (p21.1)
  - #6090: Handle empty "mark" processing instructions

CNXML-0.5 (p21)
  - Major accessibility release:
    - Added translation support for some elements, including hidden ones
    - Better automatic alt tags for images that don't have alt params
    - Changed many elements to make HTML output more semantic
    - Put table footer after tfoot so that screen readers interpret them 
      better, even though that's invalid.
    - Made solution toggler work better for keyboard and screen reader 
  - Pop-ups for terms that link to definitions in in-module glossaries.
  - Added several processing instructions as pre-0.6 hacks
  - Changed several CSS selectors to be more precise or efficient
  - Allowed glossary to have user-supplied ID.
  - #2250: Implemented more of CALS table, including: correct border 
    implementation; support of multiple tgroups; fuller support of 
    colwidth, colsep, rowsep, char, charoff, valign, and align 
    attributes (including from previous unsupported ancestors); 
    spanning via spanname attribute; better column identification.

CNXML-0.5 (p20)
  - various small changes to make the XHTML output work better with EIP 
    and be more valid in places

CNXML-0.5 (p19)
  - add translations for lens hover boxes; lenses1 merge

CNXML-0.5 (p18)
  - PDF link wording tweak to match collection printing
  - #3987: support for arbitrary Flash parameters
  - Two "forall" display fixes (#4421, #4423)

CNXML-0.5 (p17)
  - Support for display of collection subtypes

CNXML-0.5 (p16.1)
  - Fix Memorable Cases in IE

CNXML-0.5 (p16)
  - For modules in the Rice U. Press Memorable Cases collection:
    * change the "Click for Solution" and "Hide Solution" labels in

tools/check_cnxns.py
#!/usr/bin/python

import sys
import os
import codecs
from xml import sax
from lxml import etree
import lxml.sax
import urllib2
import getopt
import re
import pdb
"""
    SAX content handler to compare document and element IDs, and cnxns 
    pointing to those documents and elements, in a body of CNXML modules.
    For now we assume that the CNXML files are named 'index.cnxml' and are 
    in directories whose names are the module ID, and that these directories 
    are descendants of the first arg given to the progrem.  However, I want
    to make this program work for CNXML 0.6 as well, so I'll need to fancy up
    the opts at some point.

    I use lxml because it make use of libxml2, which in turn makes use of OASIS
    XML catalogs.  expat doesn't, and it will either retrieve all the DTDs 
    over the network for each document, or it will fail on certain valid 
    documents because it can't make sense of entity references.

    Output and reporting not finished yet.
"""

class TrackCnxnsAndIds(sax.handler.ContentHandler):

    def __init__(self):
        self.moduleId = None
        # elementIds = {
        #   <sourceModId> : {
        #     <elemId> : <elemName>
        #   }
        # }
        self.elementIds = {}
        # thisDocElementIds = {
        #   <elemId> : <elemName>
        # }
        self.thisDocElementIds = {}
        # pendingCnxns = {
        #   <targetModId> : [ (<sourceModId>, <targetElemId>) ... ]
        # }
        self.pendingCnxns = {}
        # badCnxns = {
        #   <sourceModId> : {
        #     <targetModId> : [ (<@document>, <@target>), ... ]
        #   }
        # }
        self.badCnxns = {}
        sax.handler.ContentHandler.__init__(self)

    def startDocument(self):
        self.thisDocElementIds = {}
        self.badCnxns[self.moduleId] = {}

    def endDocument(self):
        # Move elements and IDs from this doc to the accumulator
        self.elementIds[self.moduleId] = self.thisDocElementIds
        # Process any cnxns pending for this module
        for sourceModuleId, targetElementId in self.pendingCnxns.get(self.moduleId, []):
            # cnxn points to an ID, but there is no element with this ID 
            # registered for the target doc: a bad cnxn: add to badCnxns
            if targetElementId and not self.elementIds.get(self.moduleId, {}).get(targetElementId):
                self.appendToBad(sourceModuleId, self.moduleId, targetElementId)
        if self.pendingCnxns.has_key(self.moduleId):
            del self.pendingCnxns[self.moduleId]

    def startElementNS(self, name, qname, attrs):
        elementId = attrs.get((None, 'id'))
        if elementId and qname != 'document':
            self.thisDocElementIds[elementId] = name[1]
        if name == ('http://cnx.rice.edu/cnxml', 'cnxn'):
            self.handleCnxns(attrs)
        elif name == ('http://cnx.rice.edu/cnxml', 'link') and \
                attrs.get((None, 'class'), '') == 'cnxn':
            self.handleCnxns(attrs)

    def handleCnxns(self, attrs):
        document = attrs.get((None, 'document'), '').strip()
        if document:
            target_document = document
        else:
            target_document = self.moduleId
        if attrs.has_key((None, 'target')):
            target = attrs.get((None, 'target'), '').strip()
        else:
            target = attrs.get((None, 'target-id'), '').strip()
        if target:
            if self.elementIds.has_key(target_document) and not self.elementIds[target_document].has_key(target):
                self.appendToBad(self.moduleId, target_document, target)
            elif not self.elementIds.has_key(target_document):
                self.appendToPending(self.moduleId, target_document, target)
        elif document and not self.elementIds.has_key(document):
            self.appendToPending(self.moduleId, target_document, target)

    def endElementNS(self, name, qname):
        pass

    def appendToBad(self, sourceModuleId, targetModuleId, targetElementId):
        if self.badCnxns[sourceModuleId].has_key(targetModuleId):
            self.badCnxns[sourceModuleId][targetModuleId].append((targetModuleId, targetElementId))
        else:
            self.badCnxns[sourceModuleId][targetModuleId] = [(targetModuleId, targetElementId)]

    def appendToPending(self, sourceModuleId, targetModuleId, targetElementId):
        if self.pendingCnxns.has_key(targetModuleId):
            self.pendingCnxns[targetModuleId].append((sourceModuleId, targetElementId))
        else:
            self.pendingCnxns[targetModuleId] = [(sourceModuleId, targetElementId)]

    def cleanup(self, movePending=True):
        if movePending:
            for targetModuleId, targetList in self.pendingCnxns.items():
                for sourceModuleId, targetElementId in targetList:
                    self.appendToBad(sourceModuleId, targetModuleId, targetElementId)
        for k, v in self.badCnxns.items():
            if v == {}: del self.badCnxns[k]


if __name__ == '__main__':
    cnxml06_regex = re.compile(r'^(?P<moduleId>m\d+).xml$')
    optz, argz = getopt.getopt(sys.argv[1:], '6')
    cnxmlVersion = '0.5'
    if ('-6', '') in optz:
        cnxmlVersion = '0.6'
    ch = TrackCnxnsAndIds()
    start_dir = argz[0]
    i = 0
    upper = 10000
    for dpath, dnames, fnames in os.walk(start_dir):
        for fname in fnames:
            if fname == 'index.cnxml' and cnxmlVersion != '0.6':
                ch.moduleId = os.path.split(dpath)[-1]
                lxml.sax.saxify(lxml.etree.parse(os.path.join(dpath, fname)), ch)
                i += 1
                if i == upper:
                    break
            elif cnxmlVersion == '0.6':
                moduleId = None
                m = cnxml06_regex.match(fname)
                if m:
                    moduleId = m.group('moduleId')
                if moduleId:
                    ch.moduleId = moduleId
                    lxml.sax.saxify(lxml.etree.parse(os.path.join(dpath, fname)), ch)
                i += 1
                if i == upper:
                    break
        if i == upper:
            break
    ch.cleanup(movePending=True)
    sourceModuleIds = ch.badCnxns.keys()
    sourceModuleIds.sort()
    badfile = open('./bad_cnxns.csv', 'w')
    for sourceModuleId in sourceModuleIds:
        for targetModuleId, targetList in ch.badCnxns[sourceModuleId].items():
            for targetDoc, targetElementId in targetList:
                badfile.write("%s\t%s\t%s\n" % (sourceModuleId, targetDoc, targetElementId))
    badfile.close()

    elementsfile = codecs.open('./elementsids.csv', 'w', 'utf-8')
    for sourceModuleId in ch.elementIds.keys():
        for elementId, elementName in ch.elementIds[sourceModuleId].items():
            elementsfile.write("%s\t%s\t%s\n" % (sourceModuleId, elementId, elementName))
